Framework/API/src/DetectorSearcher.cpp

```diff
@@ -0,0 +1,224 @@
+
+    Eigen::Vector3d point(E1[0], E1[1], E1[2]);
+
+    // Ignore nonsensical points
+    if (point.hasNaN())
+      continue;
+
+    points.push_back(point);
+    m_indexMap.push_back(pointNo);
+  }
+
+  // create KDtree of cached detector Q vectors
+  m_detectorCacheSearch =
+      Kernel::make_unique<Kernel::NearestNeighbours<3>>(points);
+}
+
+/** Find the index of a detector given a vector in Qlab space
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+const DetectorSearcher::DetectorSearchResult
+DetectorSearcher::findDetectorIndex(const V3D &q) {
+  // quick check to see if this Q is valid
+  if (q.nullVector())
+    return std::make_tuple(false, 0);
+
+  // search using best strategy for current instrument
+  if (m_usingFullRayTrace) {
+    return searchUsingInstrumentRayTracing(q);
+  } else {
+    return searchUsingNearestNeighbours(q);
+  }
+}
+
+/** Find the index of a detector given a vector in Qlab space using a ray
+ * tracing search strategy
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+const DetectorSearcher::DetectorSearchResult
+DetectorSearcher::searchUsingInstrumentRayTracing(const V3D &q) {
+  const auto direction = convertQtoDirection(q);
+  m_rayTracer->traceFromSample(direction);
+  const auto det = m_rayTracer->getDetectorResult();
+
+  if (!det)
+    return std::make_tuple(false, 0);
+
+  const auto detIndex = m_detInfo.indexOf(det->getID());
+
+  if (m_detInfo.isMasked(detIndex) || m_detInfo.isMonitor(detIndex))
+    return std::make_tuple(false, 0);
+
+  return std::make_tuple(true, detIndex);
+}
+
+/** Find the index of a detector given a vector in Qlab space using a nearest
+ * neighbours search strategy
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+const DetectorSearcher::DetectorSearchResult
+DetectorSearcher::searchUsingNearestNeighbours(const V3D &q) {
+  const auto detectorDir = convertQtoDirection(q);
+  // find where this Q vector should intersect with "extended" space
+  const auto neighbours =
+      m_detectorCacheSearch->findNearest(Eigen::Vector3d(q[0], q[1], q[2]), 5);
+  if (neighbours.size() == 0)
+    return std::make_tuple(false, 0);
+
+  const auto result = checkInteceptWithNeighbours(detectorDir, neighbours);
+  const auto hitDetector = std::get<0>(result);
+  const auto index = std::get<1>(result);
+
+  if (hitDetector)
+    return std::make_tuple(true, m_indexMap[index]);
+
+  // Tube Gap Parameter specifically applies to CORELLI
+  if (!hitDetector && m_instrument->hasParameter("tube-gap")) {
```

Comment:
I have ideas for the future that this can be done by other means, but yes for now let's move it to another method.